ext/curl/share.c

```diff
@@ -134,6 +135,142 @@ PHP_FUNCTION(curl_share_strerror)
+ *
+ * Throws an exception if the share options are invalid.
+ */
+PHP_FUNCTION(curl_share_init_persistent)
+{
+	HashTable *share_opts = NULL;
+	zval *share_opts_entry = NULL;
+
+	zend_ulong persistent_id = 0;
+
+	php_curlsh *sh = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY_HT(share_opts)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (zend_hash_num_elements(share_opts) == 0) {
+		zend_argument_must_not_be_empty_error(1);
+		goto error;
+	}
+
+	ZEND_HASH_FOREACH_VAL(share_opts, share_opts_entry) {
+		ZVAL_DEREF(share_opts_entry);
+
+		bool failed = false;
+		zend_ulong option = zval_try_get_long(share_opts_entry, &failed);
+
+		if (failed) {
+			zend_argument_type_error(1, "must contain only int values, %s given", zend_zval_value_name(share_opts_entry));
+			goto error;
+		}
+
+		switch (option) {
+			// Disallowed options
+			case CURL_LOCK_DATA_COOKIE:
+				zend_argument_value_error(1, "must not contain CURL_LOCK_DATA_COOKIE because sharing cookies across PHP requests is unsafe");
+				goto error;
+
+			// Allowed options
+			case CURL_LOCK_DATA_DNS:
+				persistent_id |= 1 << 0;
+				break;
+			case CURL_LOCK_DATA_SSL_SESSION:
+				persistent_id |= 1 << 1;
+				break;
+			case CURL_LOCK_DATA_CONNECT:
+				persistent_id |= 1 << 2;
+				break;
+			case CURL_LOCK_DATA_PSL:
+				persistent_id |= 1 << 3;
+				break;
+
+			// Unknown options
+			default:
+				zend_argument_value_error(1, "must contain only CURL_LOCK_DATA_* constants");
+				goto error;
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	// Construct a property field for the CurlSharePersistentHandle object with the enabled share options.
+	zval share_opts_prop;
+	array_init(&share_opts_prop);
+
+	if (persistent_id & (1 << 0)) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_DNS);
+	}
+
+	if (persistent_id & (1 << 1)) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_SSL_SESSION);
+	}
+
+	if (persistent_id & (1 << 2)) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_CONNECT);
+	}
+
+	if (persistent_id & (1 << 3)) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_PSL);
+	}
+
+	object_init_ex(return_value, curl_share_persistent_ce);
+
+	zend_update_property(curl_share_persistent_ce, Z_OBJ_P(return_value), ZEND_STRL("options"), &share_opts_prop);
+	zval_ptr_dtor(&share_opts_prop);
+
+	sh = Z_CURL_SHARE_P(return_value);
+
+	zval *persisted = zend_hash_index_find(&CURL_G(persistent_curlsh), persistent_id);
+
+	// If we were able to find an existing persistent share handle, we can use it and return early.
+	if (persisted) {
+		sh->share = Z_PTR_P(persisted);
+
+		return;
+	}
+
+	// We could not find an existing share handle, so we'll have to create one.
+	sh->share = curl_share_init();
+
+	// Apply the options property to the handle. We avoid using $share_opts as zval_get_long may not necessarily return
+	// the same value as it did in the initial ZEND_HASH_FOREACH_VAL above.
+	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(&share_opts_prop), share_opts_entry) {
```

Comment:
One optional optimization you can do is using `ZEND_HASH_PACKED_FOREACH_VAL` because the `share_opts_prop` is guaranteed to be packed.